### PR TITLE
Refactor BaseHandler

### PR DESCRIPTION
**Pull request by htappen**
## Description

Refactors BaseHandler and derived classes for better consistency and reusability. Fixes #(436) and sets the stage for fixing #(436), which will be fixed in a later PR.

This pull request does include changes for the models to assume things are in batch, and updates some inconsistent properties of index_to_name.json for object detector.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Unit tests have been added for the individual handlers. I also ran example models for each of the unit tests added.

## Checklist:

- [Y] Have you added tests that prove your fix is effective or that this feature works?
- [Y] New and existing unit tests pass locally with these changes?
- [Y] Has code been commented, particularly in hard-to-understand areas?
- [Y] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: examples/image_segmenter/README.md [hunk 1]
```diff
@@ -18,90 +18,4 @@ wget https://download.pytorch.org/models/fcn_resnet101_coco-7ecb50ca.pth
     curl http://127.0.0.1:8080/predictions/fcn -T examples/image_segmenter/fcn/persons.jpg
     ```
 * Output
-
-```json
-[[[11.49452    11.49452    11.49452    ... 10.846567   10.846567
```
Comment (mycpuorg) [inline examples/image_segmenter/README.md:23]: Might be worth keeping the output of example inputs somewhere for people to cross verify. Maybe ``example_output.json``

File: examples/object_detector/index_to_name.json [hunk 2]
```diff
@@ -1,14 +1,93 @@
...
-    "kite", "baseball bat", "baseball glove", "skateboard", "surfboard", "tennis racket",
-    "bottle", "N/A", "wine glass", "cup", "fork", "knife", "spoon", "bowl",
-    "banana", "apple", "sandwich", "orange", "broccoli", "carrot", "hot dog", "pizza",
-    "donut", "cake", "chair", "couch", "potted plant", "bed", "N/A", "dining table",
-    "N/A", "N/A", "toilet", "N/A", "tv", "laptop", "mouse", "remote", "keyboard", "cell phone",
-    "microwave", "oven", "toaster", "sink", "refrigerator", "N/A", "book",
-    "clock", "vase", "scissors", "teddy bear", "hair drier", "toothbrush"
-]}
\ No newline at end of file
+{
```
Comment (mycpuorg) [inline examples/object_detector/index_to_name.json:1]: This would break existing handlers, I don't see a strong reason to do this.

File: examples/image_classifier/resnet_152_batch/resnet152_handler.py [hunk 3]
```diff
@@ -1,38 +0,0 @@
-import torch
-
```
Comment (mycpuorg) [inline examples/image_classifier/resnet_152_batch/resnet152_handler.py:2]: Deleting this example?

File: ts/torch_handler/base_handler.py [hunk 4]
```diff
@@ -69,37 +62,61 @@ def initialize(self, context):
...
     def preprocess(self, data):
         """
         Override to customize the pre-processing
-        :param data: input data
-        :return: input data
+        :param data: Python list of data items
+        :return: input tensor on a device
         """
-        return data
+        return torch.as_tensor(data, device=self.device)
```
[the remark this answers is not in the document]
Comment (htappen) [reply ts/torch_handler/base_handler.py:99]: Mostly this is to clarify what should happen so developers can consume the other methods without overriding.

File: ts/torch_handler/base_handler.py [hunk 5]
```diff
@@ -69,37 +62,61 @@ def initialize(self, context):
...
     def inference(self, data):
         """
         Override to customize the inference
-        :param data: pre-processed data
-        :return: prediction output
+        :param data: Torch tensor, matching the model input shape
+        :return: Prediction output as Torch tensor
         """
-        return self.model(data)
+        marshalled_data = data.to(self.device)
```
[the remark this answers is not in the document]
Comment (htappen) [reply ts/torch_handler/base_handler.py:107]: Yes, this is mostly to make it clear what should come out of preprocess and allow for using inference without an override.

File: ts/torch_handler/image_classifier.py [hunk 6]
```diff
@@ -17,77 +15,24 @@ class ImageClassifier(VisionHandler):
...
-
-    def __init__(self):
-        super(ImageClassifier, self).__init__()
-        self.topk = ImageClassifier.TOP_FIVE_CLASSES
+    topk = 5
+    image_processing = transforms.Compose([
+        transforms.Resize(256),
+        transforms.CenterCrop(224),
+        transforms.ToTensor(),
+        transforms.Normalize(mean=[0.485, 0.456, 0.406],
```
[the remark this answers is not in the document]
Comment (htappen) [reply ts/torch_handler/image_classifier.py:23]: Sorry, I didn't quite understand. Do you mean adding a comment?

File: examples/image_segmenter/README.md (hunk 1 above)
Comment (htappen) [reply examples/image_segmenter/README.md:23]: I can update this example. As far as I could tell, image_segmenter never actually worked correctly. So, I had to delete this big mass of text to replace it with a working example.

File: examples/object_detector/index_to_name.json (hunk 2 above)
Comment (htappen) [reply examples/object_detector/index_to_name.json:1]: The actual handler code can still handle the old case, so it's backwards compatible. This change was to bring consistency between classifiers and object detectors which had no reason as far as I could tell to have a different format for index_to_name.json

File: examples/image_classifier/resnet_152_batch/resnet152_handler.py (hunk 3 above)
Comment (htappen) [reply examples/image_classifier/resnet_152_batch/resnet152_handler.py:2]: Yes. The native handlers can now handle batching, so this extra handler is unnecessary.

File: ts/torch_handler/base_handler.py [hunk 7]
```diff
@@ -69,37 +62,61 @@ def initialize(self, context):
         # Read the mapping file, index to object name
         mapping_file_path = os.path.join(model_dir, "index_to_name.json")
 
-        if os.path.isfile(mapping_file_path):
-            with open(mapping_file_path) as f:
-                self.mapping = json.load(f)
-        else:
-            logger.warning('Missing the index_to_name.json file. Inference output will not include class name.')
+        # Load class mapping for classifiers
+        mapping_file_path = os.path.join(model_dir, "index_to_name.json")
```
Comment (dhaniram-kshirsagar) [inline ts/torch_handler/base_handler.py:66]: duplicate line.

File: ts/torch_handler/base_handler.py [hunk 8]
```diff
@@ -32,35 +30,30 @@ def initialize(self, context):
...
+        is_torchscript = True
         try:
             logger.debug('Loading torchscript model')
-            self.model = torch.jit.load(model_pt_path)
+            self.model = self._load_torchscript_model(model_pt_path)
         except RuntimeError as e:
-            # Read model definition file
+            is_torchscript = False
+
+        if not is_torchscript:
```
Comment (dhaniram-kshirsagar) [inline ts/torch_handler/base_handler.py:52]: This seems redundant. Can we remove it?

File: ts/torch_handler/base_handler.py [hunk 9]
```diff
@@ -69,37 +62,61 @@ def initialize(self, context):
...
-        :param data: input data
-        :return: input data
+        :param data: Python list of data items
+        :return: input tensor on a device
         """
-        return data
+        return torch.as_tensor(data, device=self.device)
 
-    def inference(self, data):
+    def inference(self, data, *args, **kwargs):
```
Comment (dhaniram-kshirsagar) [inline ts/torch_handler/base_handler.py:101]: At least from the `handle` method in this class, args/kwargs are not getting used. What is the use case.. via overriding handle itself?

File: examples/image_classifier/resnet_152_batch/resnet152_handler.py (hunk 3 above)
Comment (dhaniram-kshirsagar) [reply examples/image_classifier/resnet_152_batch/resnet152_handler.py:2]: Can update associated doc https://github.com/pytorch/serve/blob/master/docs/batch_inference_with_ts.md

File: ts/torch_handler/image_classifier.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dhaniram-kshirsagar) [reply ts/torch_handler/image_classifier.py:25]: Since this is same for image classifier and segmenter, can you move it as a generic method which returns this to VisionHandler and probably override it in object detector?

File: examples/image_classifier/resnet_152_batch/README.md [hunk 10]
```diff
@@ -3,7 +3,7 @@ Run the commands given in following steps from the parent directory of the root
 
 ```bash
```
Comment (dhaniram-kshirsagar) [inline examples/image_classifier/resnet_152_batch/README.md:4]: remove this as handler has been removed

File: examples/image_segmenter/README.md [hunk 11]
```diff
@@ -18,90 +18,18 @@ wget https://download.pytorch.org/models/fcn_resnet101_coco-7ecb50ca.pth
...
-  [-0.47415262 -0.47415262 -0.47415262 ... -0.4314881  -0.43148813
-   -0.43148813]
-  ...
-  [ 0.04289126  0.04289126  0.04289126 ... -0.42638034 -0.42638034
-   -0.42638034]
-  [ 0.04289126  0.04289126  0.04289126 ... -0.42638034 -0.42638034
-   -0.42638034]
-  [ 0.04289126  0.04289126  0.04289126 ... -0.42638034 -0.42638034]
-```
+[[
```
Comment (dhaniram-kshirsagar) [inline examples/image_segmenter/README.md:24]: Will the user be able to repro segmented image with this response?

File: ts/torch_handler/unit_tests/run_unit_tests.sh [hunk 12]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
```
Comment (dhaniram-kshirsagar) [inline ts/torch_handler/unit_tests/run_unit_tests.sh:1]: As far as I understood this, it will do end to end testing of default handlers without deploying them. However, it seems this will be a manual step. We are already doing end to end tests for default handlers however with deployment via `torchserve_sanity.sh`. Is there any other advantage of adding this? If yes, then we should make it part of sanity testing.

File: ts/utils/util.py [hunk 13]
```diff
@@ -21,3 +26,54 @@ def list_classes_from_module(module, parent_class=None):
...
+    for key, value in mapping.items():
+        new_value = value
+        if isinstance(new_value, list):
+            new_value = value[-1]
+        if not isinstance(new_value, str):
+            raise Exception('labels in index_to_name must be either str or [str]')
+        mapping[key] = new_value
+    return mapping
+
+def map_class_to_label(probs, mapping=None, lbl_classes=None):
```
Comment (dhaniram-kshirsagar) [inline ts/utils/util.py:58]: I see that you have modified the json files and made them consistent with respect to formats. Can you please document this somewhere so that some planning to use this util method will know what is expected.

File: ts/torch_handler/base_handler.py (hunk 7 above)
Comment (htappen) [reply ts/torch_handler/base_handler.py:66]: Removed the duplicate. Moving it to the function seems a bit odd -- most methods in Python regarding files accept a full path instead of dir + filename.

File: ts/torch_handler/base_handler.py (hunk 8 above)
Comment (htappen) [reply ts/torch_handler/base_handler.py:52]: Fixed. I'll merge in the other PR when this branch merges in.

File: examples/image_classifier/resnet_152_batch/README.md (hunk 10 above)
Comment (htappen) [reply examples/image_classifier/resnet_152_batch/README.md:4]: Sorry, I didn't quite understand that. The readme has been updated to remove the custom handler. Did you mean to remove the whole example?

File: ts/torch_handler/unit_tests/run_unit_tests.sh (hunk 12 above)
Comment (htappen) [reply ts/torch_handler/unit_tests/run_unit_tests.sh:1]: The main advantage is being able to quickly test changes on just the handlers without having to run the whole torchserve_sanity.sh script. I found it faster to iterate this way.

File: ts/utils/util.py (hunk 13 above)
Comment (htappen) [reply ts/utils/util.py:58]: Added to the docs in next commit.

File: ts/torch_handler/unit_tests/run_unit_tests.sh (hunk 12 above)
Comment (dhaniram-kshirsagar) [reply ts/torch_handler/unit_tests/run_unit_tests.sh:1]: Can we document it in  [default handler doc](https://github.com/pytorch/serve/blob/master/docs/default_handlers.md). Also, need to document batch support in default handlers [excpet text class..].

File: examples/image_classifier/resnet_152_batch/README.md (hunk 10 above)
Comment (dhaniram-kshirsagar) [reply examples/image_classifier/resnet_152_batch/README.md:4]: Please ignore. Looks good to me.

File: ts/torch_handler/unit_tests/run_unit_tests.sh (hunk 12 above)
Comment (htappen) [reply ts/torch_handler/unit_tests/run_unit_tests.sh:1]: Added to the doc.

File: examples/image_segmenter/README.md (hunk 11 above)
Comment (htappen) [reply examples/image_segmenter/README.md]: Updated this doc with the same output from the example.

**Comment by htappen** [comment]
Great, thank Dharinam!
